### PR TITLE
feat(profile): implement employment history

### PR DESCRIPTION
**Pull request by ThorsAngerVaNeT**
**🟢 Add `deploy` label if you want to deploy this Pull Request to staging environment**

#### 🧑‍⚖️ Pull Request Naming Convention

- Title should follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary)
- Do not put issue id in title
- Do not put WIP in title. Use [Draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/) functionality
- Consider to add `area:*` label(s)

* [x] I followed naming convention rules

---

#### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Performance optimization
- [ ] Refactoring
- [ ] Test Case
- [ ] Documentation update
- [ ] Other

#### 🔗 Related issue link

https://github.com/rolling-scopes/rsschool-app/issues/2202

#### 💡 Background and solution

Add employment history block to profile so students could submit in the app their working experience.
<details>

image
image
image

</details>

#### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Database migration is added or not needed
- [x] Documentation is updated/provided or not needed
- [x] Changes are tested locally

### CODE REVIEW

File: server/src/models/user.ts [hunk 1]
```diff
@@ -220,4 +220,13 @@ export class User {
...
   notificationConnections: NotificationUserConnection[] | null;
+
+  @Column({ type: Boolean, default: false })
+  jobFound: boolean;
+
+  @Column({ type: String, default: null, nullable: true })
+  jobFoundCompanyName: string | null;
+
+  @Column({ type: String, default: null, nullable: true })
+  jobFoundOfficeLocation: string | null;
```
Comment (aaliakseyenka) [inline server/src/models/user.ts:231]: I am not sure this is scalable. what are the use cases. probable we could use existing `employmentHistory` for that.   AFAIU the main purpose to mark current employment and not the fact that student found the job at certain point.

Comment (ThorsAngerVaNeT) [reply server/src/models/user.ts:231]: > I am not sure this is scalable. what are the use cases. probable we could use existing `employmentHistory` for that. AFAIU the main purpose to mark current employment and not the fact that student found the job at certain point.

I understand this oppositely, but you could be right.
@valerydluski could you please specify this task more precisely or comment on Artem's suggestion?

Comment (valerydluski) [reply server/src/models/user.ts:231]: @ThorsAngerVaNeT 
you can make changes without my participation. I am not a business analyst

File: client/src/components/Profile/JobFoundButtonWithModal.tsx [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+      }
+      await profileApi.updateJobFound(values);
+      setSubmitted(true);
+    } catch (e) {
+      let error = 'Unknown error';
+      if (e instanceof AxiosError) {
+        error = e.response?.data?.message;
+      } else if (e instanceof Error) {
+        error = e.message;
+      }
```
Comment (apalchys) [inline client/src/components/Profile/JobFoundButtonWithModal.tsx:45]: It's better to extract to reusable function

File: server/src/models/user.ts (hunk 1 above)
Comment (apalchys) [reply server/src/models/user.ts:231]: @ThorsAngerVaNeT Lets use `employmentHistory` and save information there.  I don't think we use it now anywhere.
So, please collect title, company, `dateFrom` and mark as `toPresent:true`

Comment (apalchys) [reply server/src/models/user.ts:231]: You may also extend and add `locationName` to there and use the same that we use for searching location during regstration.

Comment (ThorsAngerVaNeT) [reply server/src/models/user.ts:231]: > @ThorsAngerVaNeT Lets use `employmentHistory` and save information there. I don't think we use it now anywhere. So, please collect title, company, `dateFrom` and mark as `toPresent:true`

Sure, I was going to use `employmentHistory` and redo this task, but I was working on other tasks

File: client/src/components/Profile/EmploymentCard.tsx [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    const employmentHistory = employments;
+    const isUpdated = await updateProfile({
+      employmentHistory: employmentHistory.map(employmentRecordFormItemToDto),
+    });
+
+    if (!isUpdated) {
+      return;
+    }
+
+    setDisplayEmployments(employments);
```
Comment (I-vasilich-I) [inline client/src/components/Profile/EmploymentCard.tsx:87]: related to `EmploymentHistoryDisplayItem` comment
So after save, if I don't reload the page I won't see employment from-present-duration, because it relies on `dateTo` to be truthy, 

this could solve it:
```js
setDisplayEmployments(employments.map(( item ) => ({...item, dateTo: item.dateTo ? dayjs(item.dateTo) : dayjs() })));
```

but IMHO the way you set `dateTo` on `employmentRecordDtoToFormItem` seams not right, if it's not set, it should stay not set
image 
image

better to handle it on `EmploymentHistoryDisplayItem`

File: client/src/components/Profile/EmploymentCard.tsx [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+      saveProfile={handleSave}
+      cancelChanges={handleCancel}
+      isSaveDisabled={isSaveDisabled}
+      content={
+        displayEmployments.length ? (
+          <List itemLayout="horizontal" dataSource={displayEmployments} renderItem={renderContentItem} />
+        ) : null
+      }
+      profileSettingsContent={
+        <>
```
[the remark this answers is not in the document]
Comment (ThorsAngerVaNeT) [reply client/src/components/Profile/EmploymentCard.tsx:116]: You're right. Thanks

File: client/src/components/Profile/EmploymentCard.tsx (hunk 3 above)
Comment (ThorsAngerVaNeT) [reply client/src/components/Profile/EmploymentCard.tsx:87]: > this could solve it:
> 
> ```js
> setDisplayEmployments(employments.map(( item ) => ({...item, dateTo: item.dateTo ? dayjs(item.dateTo) : dayjs() })));
> ```

Yeah, I had such solution before somewhere, but for some reason I get rid of it.
Thanks I will double check this issue

File: client/src/components/Profile/EmploymentHistory/EmploymentHistoryFormItem.tsx [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            dateTo,
+            toPresent: isToPresent,
+            companyName,
+          }}
+        />
+        <Form.Item
+          {...restField}
+          name={[name, 'title']}
+          label="Position title"
+          rules={[{ required: true, message: '${label} is required' }]}
```
[the remark this answers is not in the document]
Comment (ThorsAngerVaNeT) [reply client/src/components/Profile/EmploymentHistory/EmploymentHistoryFormItem.tsx:50]: Yes, it's **antd** [validateMessages](https://ant.design/components/form#validatemessages)

File: client/src/components/Profile/MainCard.tsx [hunk 6]
```diff
@@ -120,34 +120,36 @@ const MainCard = ({ data, isEditingModeEnabled, updateProfile }: Props) => {
...
-                      style={{ flex: 1 }}
-                      onChange={handleLocationChange}
-                      location={locationSelectValue}
-                    />
-                  </Row>
-                </Col>
-              </Row>
-            }
-          />
+          <>
```
[the remark this answers is not in the document]
Comment (ThorsAngerVaNeT) [reply client/src/components/Profile/MainCard.tsx:123]: Nice catch!

File: client/src/components/Profile/EmploymentCard.tsx [hunk 7]
```diff
@@ -0,0 +1,151 @@
+import React, { useEffect, useMemo, useState } from 'react';
```
Comment (valerydluski) [inline client/src/components/Profile/EmploymentCard.tsx:1]: do you need React import here?

Comment (ThorsAngerVaNeT) [reply client/src/components/Profile/EmploymentCard.tsx:1]: Nope, thank you.

File: client/src/api/api.ts [hunk 8]
```diff
@@ -2555,6 +2555,49 @@ export interface Education {
...
+     * @type {boolean}
+     * @memberof EmploymentRecordDto
+     */
+    'toPresent': boolean;
+    /**
+     * 
+     * @type {object}
+     * @memberof EmploymentRecordDto
+     */
+    'officeLocation'?: object;
```
Comment (aaliakseyenka) [inline client/src/api/api.ts:2599]: should it be object here?

File: client/src/components/Profile/EmploymentCard.tsx [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+  const [form] = Form.useForm<{ employmentHistory: EmploymentRecordFormItem[] }>();
+  const values = Form.useWatch([], form);
+
+  const employmentRecordFormItemToDto = (employment: EmploymentRecordFormItem): EmploymentRecordDto => ({
+    ...employment,
+    dateFrom: employment.dateFrom.format(dateFormat),
+    dateTo: employment.toPresent ? '' : employment.dateTo.format(dateFormat),
+  });
+
+  const employmentRecordDtoToFormItem = (employment: EmploymentRecordDto): EmploymentRecordFormItem => ({
```
Comment (aaliakseyenka) [inline client/src/components/Profile/EmploymentCard.tsx:43]: do you need this functions to be defined inside component?

File: client/src/components/Profile/EmploymentCard.tsx [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+    ...employment,
+    dateFrom: employment.dateFrom ? dayjs(employment.dateFrom) : dayjs(),
+    dateTo: employment.dateTo ? dayjs(employment.dateTo) : dayjs(),
+  });
+
+  const [displayEmployments, setDisplayEmployments] = useState<EmploymentRecordFormItem[]>(
+    data.map(employmentRecordDtoToFormItem),
+  );
+  const [employments, setEmployments] = useState<EmploymentRecordFormItem[]>(displayEmployments);
```
Comment (aaliakseyenka) [inline client/src/components/Profile/EmploymentCard.tsx:52]: IMO looks weird to have 2 states for the same data

File: common/models/profile.ts [hunk 11]
```diff
@@ -176,3 +176,9 @@ export type LegacyFeedback = {
     dataStructures: number;
   };
 };
+
+export interface JobFoundInfo {
+  jobFound: boolean;
+  jobFoundCompanyName?: string | null;
+  jobFoundOfficeLocation?: string | null;
```
Comment (aaliakseyenka) [inline common/models/profile.ts:183]: do you really need to repeat this `jobFound` in each property?

File: client/src/components/Profile/EmploymentHistory/EmploymentHistoryFormItem.tsx [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+type EmploymentHistoryFormItemPros = {
+  name: number;
+  restField: { fieldKey?: number | undefined };
+  remove: (index: number | number[]) => void;
+  form: FormInstance<{ employmentHistory: EmploymentRecordFormItem[] }>;
+};
+
+const EmploymentHistoryFormItem = ({ name, restField, remove, form }: EmploymentHistoryFormItemPros) => {
+  const { title, dateFrom, dateTo, toPresent, companyName, officeLocation } =
+    form.getFieldValue(['employmentHistory', name]) ?? {};
```
Comment (aaliakseyenka) [inline client/src/components/Profile/EmploymentHistory/EmploymentHistoryFormItem.tsx:19]: why do you need explicitly get this data rather than us form items to consume it automatically?

File: client/src/components/Profile/EmploymentHistory/EmploymentHistoryFormItem.tsx [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+  const [isToPresent, setToPresent] = useState<boolean>(toPresent);
+  const [isDateToRequired, setIsDateToRequired] = useState<boolean>(!toPresent);
+  const [locationSelectValue, setLocationSelectValue] = useState(officeLocation);
```
Comment (aaliakseyenka) [inline client/src/components/Profile/EmploymentHistory/EmploymentHistoryFormItem.tsx:22]: could u. please clarify why do you need state for them. why antd form state is not sufficient?

File: client/src/components/Profile/EmploymentHistory/EmploymentHistoryDisplayItem.tsx [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+}: {
+  employmentRecord: Omit<EmploymentRecordFormItem, 'officeLocation'>;
+}) => {
+  const { title, dateFrom, dateTo, toPresent, companyName } = employmentRecord;
+
+  if (!title && !companyName) {
+    return <Title level={5}>(Empty)</Title>;
+  }
+
+  const getWorkingPeriod = (dateFrom: Dayjs, dateTo: Dayjs): string => {
```
Comment (aaliakseyenka) [inline client/src/components/Profile/EmploymentHistory/EmploymentHistoryDisplayItem.tsx:19]: plz move out the functions that are not related to rendering out of component body in all places

Comment (ThorsAngerVaNeT) [reply client/src/components/Profile/EmploymentHistory/EmploymentHistoryDisplayItem.tsx:19]: Ok, thanks

File: client/src/components/Profile/EmploymentCard.tsx (hunk 9 above)
Comment (ThorsAngerVaNeT) [reply client/src/components/Profile/EmploymentCard.tsx:43]: Not really, thank you

File: nestjs/src/profile/profile.service.ts [hunk 15]
```diff
@@ -209,6 +211,11 @@ export class ProfileService {
     return this.userRepository.findOneOrFail({ where: { githubId }, relations: ['students'] });
   }
 
+  public async getEmploymentHistory(userId: number) {
+    const user = await this.userRepository.findOneOrFail({ where: { id: userId } });
```
[the remark this answers is not in the document]
Comment (ThorsAngerVaNeT) [reply nestjs/src/profile/profile.service.ts:215]: Good suggestion, thank you
